src/v2i-hub/docs/Programming_Guide.md

```diff
@@ -0,0 +1,1233 @@
+# Programming Guide
+
+The V2X Hub Programming Guide was developed during the V2I Reference Implementation project for developers of plugins for the V2X Hub in supporting Connected Vehicle (CV) technology. This programming guide gives details on how to create a plugin for the V2X Hub based on the Example Plugin created by Battelle. The Example Plugin can be obtained from OSADP, along with the other V2X Hub software that you will need to compile and setup to do your own V2X Hub development.
+
+The V2X Hub consists of software modules written in C and C++ for Linux. The developer wanting to create a plugin for the V2X Hub must be familiar with C, C++ and Linux, and have a basic knowledge of object-oriented design. Some other important software development practices to be familiar with are threading, network communications, external library integration, and data serialization just to name a few. Development is done using the following tools: cmake, make, g++ compiler, and optionally eclipse. Developers should have familiarity with those tools to successfully create a V2X Hub plugin.
+
+Additional information on the V2X Hub can be found in the following documents. It is suggested to start with the V2X Hub Guidebook.
+
+- Integrated Vehicle-to-Infrastructure Prototype (IVP) V2X Hub Design, March 2017
+- Integrated Vehicle-to-Infrastructure Prototype (IVP) V2X Hub Interface Control Document (ICD), March 2017
+- V2X Hub Guidebook
+- V2X Hub Plugins
+- V2X Hub Deployment Guide
+- V2X Hub Software Configuration Guide
+- V2X Hub Administration Portal User Guide
+
+## Example Plugin
+
+The Example Plugin showcases the basic functionality needed by a V2X Hub plugin. The plugin has simple API calls to create, route, and decode messages from the V2X Hub Core. Examples for setting status values, retrieving configuration values and writing logs are also in the example. The Example Plugin is a good foundation to modify when creating our own plugin for the V2X Hub. See [SampleBSMPlugin](../../../examples/tmx-exampleapps/SampleBSMPlugin/) for example plugin
+
+### Plugin Structure
+
+The Example Plugin structure is as follows:
+
+- Example Plugin
+  - src
+    - SampleBSMPlugin.cpp
+    - SampleBSMPlugin.h
+  - CMakeLists.txt
+  - manifest.json
+  - README.md
+
+The main source code for the plugin resides in the ExamplePlugin.cpp file. This C++ class extends the PluginClient class from the V2X Hub API. The PluginClient has methods that can be overwritten to retrieve settings and messages. The next section will describe the source code needed to do the basic functionality for a plugin, and chapter 2 will walk through the process of creating a plugin based on the Example Plugin. The SampleData files are simply examples on how to include other C++ source files into a V2X Hub Plugin. They are currently not used.
```

Comment:
PluginClient or ClockAwarePlugin?

Reply:
Not seeing SampleData files?

Reply:
Its refering to the ExamplePlugin SampleData.cpp. I will correct the part of this referring to the wrong base class but I want to save correction related to the example plugin for a story in the Update Documentation epic to revise/consolidate our example/tutorial plugin.